### PR TITLE
[SPARK-20114][ML][FOLLOW-UP] spark.ml parity for sequential pattern mining - PrefixSpan

### PR DESCRIPTION
**Pull request by WeichenXu123**
## What changes were proposed in this pull request?

Change `PrefixSpan` into a class with param setter/getters.
This address issues mentioned here:
https://github.com/apache/spark/pull/20973#discussion_r186931806

## How was this patch tested?

UT.

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

**Comment by WeichenXu123** [comment]
@mengxr @jkbradley

File: mllib/src/main/scala/org/apache/spark/ml/fpm/PrefixSpan.scala [hunk 1]
```diff
@@ -35,7 +37,87 @@ import org.apache.spark.sql.types.{ArrayType, LongType, StructField, StructType}
...
+  /**
+   * the minimal support level of the sequential pattern, any pattern that
+   * appears more than (minSupport * size-of-the-dataset) times will be output
+   * (default value: `0.1`).
+   * @group param
+   */
+  @Since("2.4.0")
+  val minSupport = new DoubleParam(this, "minSupport", "the minimal support level of the " +
+    "sequential pattern, any pattern that appears more than (minSupport * size-of-the-dataset) " +
+    "times will be output", ParamValidators.gt(0.0))
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/fpm/PrefixSpan.scala]: gt -> gtEq

File: mllib/src/main/scala/org/apache/spark/ml/fpm/PrefixSpan.scala [hunk 2]
```diff
@@ -35,7 +37,87 @@ import org.apache.spark.sql.types.{ArrayType, LongType, StructField, StructType}
  */
 @Since("2.4.0")
 @Experimental
-object PrefixSpan {
+final class PrefixSpan(@Since("2.4.0") override val uid: String) extends Params {
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/fpm/PrefixSpan.scala]: In the doc, mention that this class is not yet an Estimator/Transformer and link to `findFrequentSequentialPatterns` method.

File: mllib/src/main/scala/org/apache/spark/ml/fpm/PrefixSpan.scala [hunk 3]
```diff
@@ -35,7 +37,87 @@ import org.apache.spark.sql.types.{ArrayType, LongType, StructField, StructType}
...
+
+  @Since("2.4.0")
+  def this() = this(Identifiable.randomUID("prefixSpan"))
+
+  /**
+   * the minimal support level of the sequential pattern, any pattern that
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/fpm/PrefixSpan.scala]: Use uppercase for the first char:

"""
Param for the minimal support level (default: `0.1`).
Sequential patterns that appear more than (minSupport * size-of-the-dataset) times are identified as frequent sequential patterns.
"""

File: mllib/src/main/scala/org/apache/spark/ml/fpm/PrefixSpan.scala [hunk 4]
```diff
@@ -35,7 +37,87 @@ import org.apache.spark.sql.types.{ArrayType, LongType, StructField, StructType}
...
+    "sequential pattern, any pattern that appears more than (minSupport * size-of-the-dataset) " +
+    "times will be output", ParamValidators.gt(0.0))
+
+  /** @group getParam */
+  @Since("2.4.0")
+  def getMinSupport: Double = $(minSupport)
+
+  /**
+   * Set the minSupport parameter.
+   * Default is 1.0.
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/fpm/PrefixSpan.scala]: The default is wrong. We don't need doc for the setters and getters. Just leave "@group setParam".

File: mllib/src/main/scala/org/apache/spark/ml/fpm/PrefixSpan.scala [hunk 5]
```diff
@@ -35,7 +37,87 @@ import org.apache.spark.sql.types.{ArrayType, LongType, StructField, StructType}
...
+   *
+   * @group setParam
+   */
+  @Since("1.3.0")
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/fpm/PrefixSpan.scala]: This is also wrong. The method is new.

File: mllib/src/main/scala/org/apache/spark/ml/fpm/PrefixSpan.scala [hunk 6]
```diff
@@ -35,7 +37,87 @@ import org.apache.spark.sql.types.{ArrayType, LongType, StructField, StructType}
...
+  def setMinSupport(value: Double): this.type = set(minSupport, value)
+
+  /**
+   * the maximal length of the sequential pattern
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/fpm/PrefixSpan.scala]:
~~~
Param for the maximal pattern length (default: `10`).
~~~

Just copy the doc from mllib.fpm.PrefixSpan.

File: mllib/src/main/scala/org/apache/spark/ml/fpm/PrefixSpan.scala [hunk 7]
```diff
@@ -35,7 +37,87 @@ import org.apache.spark.sql.types.{ArrayType, LongType, StructField, StructType}
...
+   * @group param
+   */
+  @Since("2.4.0")
+  val maxPatternLength = new IntParam(this, "maxPatternLength",
+    "the maximal length of the sequential pattern",
+    ParamValidators.gt(0))
+
+  /** @group getParam */
+  @Since("2.4.0")
+  def getMaxPatternLength: Double = $(maxPatternLength)
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/fpm/PrefixSpan.scala]: return `Int`

File: mllib/src/main/scala/org/apache/spark/ml/fpm/PrefixSpan.scala [hunk 8]
```diff
@@ -35,7 +37,87 @@ import org.apache.spark.sql.types.{ArrayType, LongType, StructField, StructType}
...
+   * Set the maxPatternLength parameter.
+   * Default is 10.
+   *
+   * @group setParam
+   */
+  @Since("2.4.0")
+  def setMaxPatternLength(value: Int): this.type = set(maxPatternLength, value)
+
+  /**
+   * The maximum number of items (including delimiters used in the
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/fpm/PrefixSpan.scala]: Same here. Just copy the doc from `mllib.fpm.PrefixSpan`.

File: mllib/src/main/scala/org/apache/spark/ml/fpm/PrefixSpan.scala [hunk 9]
```diff
@@ -35,7 +37,87 @@ import org.apache.spark.sql.types.{ArrayType, LongType, StructField, StructType}
...
+  @Since("2.4.0")
+  val maxLocalProjDBSize = new LongParam(this, "maxLocalProjDBSize",
+    "The maximum number of items (including delimiters used in the internal storage format) " +
+    "allowed in a projected database before local processing. If a projected database exceeds " +
+     "this size, another iteration of distributed prefix growth is run",
+    ParamValidators.gt(0))
+
+  /** @group getParam */
+  @Since("2.4.0")
+  def getMaxLocalProjDBSize: Double = $(maxLocalProjDBSize)
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/fpm/PrefixSpan.scala]: Long

File: mllib/src/main/scala/org/apache/spark/ml/fpm/PrefixSpan.scala [hunk 10]
```diff
@@ -63,26 +135,22 @@ object PrefixSpan {
   @Since("2.4.0")
   def findFrequentSequentialPatterns(
       dataset: Dataset[_],
-      sequenceCol: String,
-      minSupport: Double,
-      maxPatternLength: Int,
-      maxLocalProjDBSize: Long): DataFrame = {
+      sequenceCol: String): DataFrame = {
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/fpm/PrefixSpan.scala]: Why not making it a param?

**Review by mengxr — CHANGES_REQUESTED** [review]

**Pull request closed without merging**